src/lib/validationRules/remoteDocumentAsJsonLd/remoteDocumentAsJsonLd.ts

```diff
@@ -0,0 +1,191 @@
+//
+// Copyright 2022 Inrupt Inc.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal in
+// the Software without restriction, including without limitation the rights to use,
+// copy, modify, merge, publish, distribute, sublicense, and/or sell copies of the
+// Software, and to permit persons to whom the Software is furnished to do so,
+// subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED,
+// INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A
+// PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT
+// HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION
+// OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE
+// SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
+//
+
+// eslint-disable-next-line no-shadow
+import { fetch, request, Response } from "undici";
+import {
+  ValidationRule,
+  RuleResult,
+  ValidationContext,
+  OIDC_CONTEXT,
+  ClientIdDocument,
+} from "../../types";
+
+const remoteDocumentAsJsonLd: ValidationRule = {
+  rule: {
+    type: "remote",
+    name: "Remote Document must be a correct ld+json document",
+    description:
+      "The Client Identifier Document must contain the correct Solid OIDC context. The server must serve a mime type of `application/json` or `application/ld+json`",
+  },
+  check: async (context: ValidationContext) => {
```

Comment:
A potentially sneaky thing we can do here, especially as we control the order of execution is to modify the context value, e.g.,

```
const validationContext = createValidationContext({ documentIri: "...", document: null });

const remoteResults = await performRemoteValidation(validationContext);

const documentResults = await performLocalValidation(validationContext);

return [ ...remoteResults, ...documentResults ]
```

Where in `performLocalValidation` it short-circuits the other checks if `validationContext.document` is null ("No document contents supplied, this may be because we were unable to fetch your Client Identifier Document")

Then in `remoteDocumentValidation` (perhaps a better name for this rule?) once we've successfully validated or fetched the document, you do `context.document = fetchedDocument`; then the subsequent rules just _work_

Reply:
Mhh, somehow I like the idea, on the other hand, it feels like a bad sin to me to use implicit state like that.. I'll leave it with a more explicit workflow for now.